File: app/src/main/java/com/example/ygg/personnalisation_fragment_v4/MainActivity.java
package com.example.ygg.personnalisation_fragment_v4;

import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.ColorDrawable;
import android.support.constraint.ConstraintLayout;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.AppCompatCheckBox;
import android.support.v7.widget.SwitchCompat;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.SeekBar;
import android.widget.TextView;

import com.madrapps.pikolo.HSLColorPicker;
import com.madrapps.pikolo.listeners.SimpleColorSelectionListener;

import java.lang.reflect.Array;
import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        final HSLColorPicker colorPicker = (HSLColorPicker) findViewById(R.id.colorPickerPikolo);
        colorPicker.setColorSelectionListener(new SimpleColorSelectionListener() {
            @Override
            public void onColorSelected(int color) {


                SeekBar sb = (SeekBar) findViewById(R.id.seekbar);
                Toolbar actionBarToolbar = (Toolbar)findViewById(R.id.action_bar);
                View bg = findViewById(R.id.mainBG);
                SwitchCompat switchCompat = (SwitchCompat) findViewById(R.id.switchControl);


                int[][] states = new int[][] {
                        new int[] {-android.R.attr.state_checked},
                        new int[] {android.R.attr.state_checked},
                };

                int oposite = getContrastColor(color);



                // ******************   T H E M E S   ****************** //

                // ***** DARK *****

                    int[] thumbColors = new int[] {
                            Color.WHITE,
                            lighten(oposite, 0.6f)
                    };

                    int[] trackColors = new int[] {
                            Color.parseColor("#C0C0C0"),
                            oposite
                    };

                    int extrude = lighten(color, 0.6f);
                    int exO = lighten(oposite, 0.4f);

                    int txt = Color.WHITE;
                    int title = darken(color, 0.4f);


                // ***** LIGHT *****
                if( Color.luminance(color) > 0.2f ) {

                    thumbColors = new int[] {
                            Color.BLACK,
                            darken(color, 0.6f)
                    };

                    trackColors = new int[] {
                            Color.GRAY,
                            lighten(color, 0.6f)
                    };

                    extrude = darken(color, 0.8f);
                    exO = darken(oposite, 0.6f);

                    txt = darken(color, 0.4f);
                    title = Color.WHITE;
                }
                // ****************** //

                // ***************************************************** //



                // BACKGROUND
                getSupportActionBar().setBackgroundDrawable( new ColorDrawable( oposite ));
                bg.setBackgroundColor( color );

                // SEPARATE
                findViewById(R.id.s1).setBackgroundColor(extrude);
                findViewById(R.id.s2).setBackgroundColor(extrude);
                findViewById(R.id.s3).setBackgroundColor(extrude);

                // SEEKBAR
                sb.getProgressDrawable().setColorFilter( exO, PorterDuff.Mode.MULTIPLY);
                sb.getThumb().setColorFilter( oposite, PorterDuff.Mode.SRC_ATOP);

                // SWITCH
                DrawableCompat.setTintList(DrawableCompat.wrap(switchCompat.getThumbDrawable()), new ColorStateList(states, thumbColors));
                DrawableCompat.setTintList(DrawableCompat.wrap(switchCompat.getTrackDrawable()), new ColorStateList(states, trackColors));

                // EDIT TEXT
                ((EditText)findViewById(R.id.editText)).setTextColor(color);

                // TEXT COLOR
                ((TextView)findViewById(R.id.label1)).setTextColor(txt);
                ((TextView)findViewById(R.id.label2)).setTextColor(txt);
                ((TextView)findViewById(R.id.label3)).setTextColor(txt);

                if (actionBarToolbar != null) { actionBarToolbar.setTitleTextColor(title); }







//                // TEST ****
//                int test = lighten(color, 0.8f);

//                View test = findViewById(R.id.mainBG).findViewWithTag("mon_tag");
//
//                if(test != null) { test.setBackgroundColor(lighten(color, 0.8f)); }



//                ArrayList<View> result = getViewsByTag( (ViewGroup) findViewById(R.id.mainBG), "mon_tag" );
//
//                if(!result.isEmpty()) {
//                    View separate = result.get(0);
//
//                    if (separate != null) {
//                        separate.setBackgroundColor(lighten(color, 0.8f));
//                    }
//                }
            }
        });

    }

    public static int getContrastColor(int color) {

        float r = 128 + ( 128 - Color.red(color) ),
                g = 128 + ( 128 - Color.green(color) ),
                b = 128 + ( 128 - Color.blue(color) );

        return Color.rgb( (int)r, (int)g, (int)b );
    }

    public static int darken(int color, float percent) {

        float[] hsv = { 0f, 0f, 0f };
        Color.colorToHSV(color, hsv);

        hsv[2] *= percent;

        return Color.HSVToColor(hsv);
    }

    public static int lighten(int color, float percent) {

        float[] hsv = { 0f, 0f, 0f };
        Color.colorToHSV(color, hsv);

        hsv[2] /= percent;

        return Color.HSVToColor(hsv);
    }



//    public static ArrayList<View> getViewsByTag(ViewGroup root, String tag){
//        ArrayList<View> views = new ArrayList<View>();
//        final int childCount = root.getChildCount();
//        for (int i = 0; i < childCount; i++) {
//            final View child = root.getChildAt(i);
//            if (child instanceof ViewGroup) {
//                views.addAll(getViewsByTag((ViewGroup) child, tag));
//            }
//
//            final Object tagObj = child.getTag();
//            if (tagObj != null && tagObj.equals(tag)) {
//                views.add(child);
//            }
//
//        }
//        return views;
//    }
}

File: app/src/main/java/com/example/ygg/personnalisation_fragment_v4/MainActivity.kt
package com.example.ygg.personnalisation_fragment_v4

import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.ColorDrawable
import android.support.constraint.ConstraintLayout
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.AppCompatCheckBox
import android.support.v7.widget.SwitchCompat
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.SeekBar
import android.widget.TextView

// antonioleiva PACKAGES ->
import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.text.Editable
import android.text.TextWatcher
import android.widget.LinearLayout
import kotlin.reflect.KProperty


import com.madrapps.pikolo.HSLColorPicker
import com.madrapps.pikolo.listeners.SimpleColorSelectionListener

import java.lang.reflect.Array
import java.util.ArrayList

class MainActivity : AppCompatActivity() {

    // PREFERENCES
    private var textValue: String by DelegatesExt.preference(this, "textValue", "Pascal Ladale \uD83D\uDC35")
    private var switchValue: Boolean by DelegatesExt.preference(this, "switchValue", true)
    private var seekValue: Int by DelegatesExt.preference(this, "seekValue", 280)
    private var colorValue: Int by DelegatesExt.preference(this, "colorValue", Color.WHITE)




    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val et = findViewById(R.id.editText) as TextView
        val sb = findViewById(R.id.seekbar) as SeekBar
        val actionBarToolbar = findViewById(R.id.action_bar) as Toolbar
        val bg = findViewById(R.id.mainBG) as LinearLayout
        val switchCompat = findViewById(R.id.switchControl) as SwitchCompat

        Log.d("mes_logs", colorValue.toString())

        ChangeColor(colorValue)
        et.setText(textValue)
        switchCompat.setChecked(switchValue)
        sb.setProgress(seekValue, true)



        et.addTextChangedListener(object : TextWatcher {
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) { }
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) { }
            override fun afterTextChanged(s: Editable) { textValue = s.toString() }
        })



        switchCompat.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                if(switchValue) { switchValue = false }
                else { switchValue = true }
            }
        })


        sb.setOnSeekBarChangeListener(SeekBarListener());



         // C O L O R   P I C K E R
        val colorPicker = findViewById(R.id.colorPickerPikolo) as HSLColorPicker
        colorPicker.setColorSelectionListener(object : SimpleColorSelectionListener() {
            override fun onColorSelected(color: Int) {

                colorValue = color

                ChangeColor(color)

            }
        })

    }



        fun ChangeColor(color: Int) {

        val et = findViewById(R.id.editText) as TextView
        val sb = findViewById(R.id.seekbar) as SeekBar
        val actionBarToolbar = findViewById(R.id.action_bar) as Toolbar
        val bg = findViewById(R.id.mainBG) as LinearLayout
        val switchCompat = findViewById(R.id.switchControl) as SwitchCompat


        val states = arrayOf(intArrayOf(-android.R.attr.state_checked), intArrayOf(android.R.attr.state_checked))

        val oposite = getContrastColor(color)


        // ******************   T H E M E S   ****************** //

        // ***** DARK *****

        var thumbColors = intArrayOf(Color.WHITE, lighten(oposite, 0.6f))

        var trackColors = intArrayOf(Color.parseColor("#C0C0C0"), oposite)

        var extrude = lighten(color, 0.6f)
        var exO = lighten(oposite, 0.4f)

        var txt = Color.WHITE
        var title = darken(color, 0.4f)


        // ***** LIGHT *****
        if (Color.luminance(color) > 0.2f) {

            thumbColors = intArrayOf(Color.BLACK, darken(color, 0.6f))

            trackColors = intArrayOf(Color.GRAY, lighten(color, 0.6f))

            extrude = darken(color, 0.8f)
            exO = darken(oposite, 0.6f)

            txt = darken(color, 0.4f)
            title = Color.WHITE
        }
        // ****************** //

        // ***************************************************** //


        // BACKGROUND
        supportActionBar!!.setBackgroundDrawable(ColorDrawable(oposite))
        bg.setBackgroundColor(color)

        // SEPARATE
        findViewById(R.id.s1).setBackgroundColor(extrude)
        findViewById(R.id.s2).setBackgroundColor(extrude)
        findViewById(R.id.s3).setBackgroundColor(extrude)

        // SEEKBAR
        sb.progressDrawable.setColorFilter(exO, PorterDuff.Mode.MULTIPLY)
        sb.thumb.setColorFilter(oposite, PorterDuff.Mode.SRC_ATOP)

        // SWITCH
        DrawableCompat.setTintList(DrawableCompat.wrap(switchCompat.thumbDrawable), ColorStateList(states, thumbColors))
        DrawableCompat.setTintList(DrawableCompat.wrap(switchCompat.trackDrawable), ColorStateList(states, trackColors))

        // EDIT TEXT
        et.setTextColor(color)

        // TEXT COLOR
        (findViewById(R.id.label1) as TextView).setTextColor(txt)
        (findViewById(R.id.label2) as TextView).setTextColor(txt)
        (findViewById(R.id.label3) as TextView).setTextColor(txt)

        actionBarToolbar.setTitleTextColor(title)

    }



    companion object {

        fun getContrastColor(color: Int): Int {

            val r = (128 + (128 - Color.red(color))).toFloat()
            val g = (128 + (128 - Color.green(color))).toFloat()
            val b = (128 + (128 - Color.blue(color))).toFloat()

            return Color.rgb(r.toInt(), g.toInt(), b.toInt())
        }

        fun darken(color: Int, percent: Float): Int {

            val hsv = floatArrayOf(0f, 0f, 0f)
            Color.colorToHSV(color, hsv)

            hsv[2] *= percent

            return Color.HSVToColor(hsv)
        }

        fun lighten(color: Int, percent: Float): Int {

            val hsv = floatArrayOf(0f, 0f, 0f)
            Color.colorToHSV(color, hsv)

            hsv[2] /= percent

            return Color.HSVToColor(hsv)
        }
    }



    inner class SeekBarListener : SeekBar.OnSeekBarChangeListener {

        override fun onProgressChanged(seekBar: SeekBar, progress: Int,
                                       fromUser: Boolean) {
            seekValue = progress
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {}

        override fun onStopTrackingTouch(seekBar: SeekBar) {}

    }
}




// antonioleiva ->

object DelegatesExt {
    fun <T> notNullSingleValue() = NotNullSingleValueVar<T>()
    fun <T> preference(context: Context, name: String,
                       default: T) = Preference(context, name, default)
}

class NotNullSingleValueVar<T> {

    private var value: T? = null

    operator fun getValue(thisRef: Any?, property: KProperty<*>): T =
            value ?: throw IllegalStateException("${property.name} not initialized")

    operator fun setValue(thisRef: Any?, property: KProperty<*>, value: T) {
        this.value = if (this.value == null) value
        else throw IllegalStateException("${property.name} already initialized")
    }
}

class Preference<T>(private val context: Context, private val name: String,
                    private val default: T) {

    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences("default", Context.MODE_PRIVATE)
    }

    operator fun getValue(thisRef: Any?, property: KProperty<*>): T = findPreference(name, default)

    operator fun setValue(thisRef: Any?, property: KProperty<*>, value: T) {
        putPreference(name, value)
    }

    @Suppress("UNCHECKED_CAST")
    private fun findPreference(name: String, default: T): T = with(prefs) {
        val res: Any = when (default) {
            is Long -> getLong(name, default)
            is String -> getString(name, default)
            is Int -> getInt(name, default)
            is Boolean -> getBoolean(name, default)
            is Float -> getFloat(name, default)
            else -> throw IllegalArgumentException("This type can be saved into Preferences")
        }

        res as T
    }

    @SuppressLint("CommitPrefEdits")
    private fun putPreference(name: String, value: T) = with(prefs.edit()) {
        when (value) {
            is Long -> putLong(name, value)
            is String -> putString(name, value)
            is Int -> putInt(name, value)
            is Boolean -> putBoolean(name, value)
            is Float -> putFloat(name, value)
            else -> throw IllegalArgumentException("This type can't be saved into Preferences")
        }.apply()
    }
}
